### PR TITLE
feat:Allowed updating of events

### LINKED ISSUE #936: Add an option to edit an event
### Feature Request

**Feature Description**
There should be an option to edit an already created event.

**Would you like to work on the issue?**
Yes

### PR DESCRIPTION
**Pull request by sridharjajoo**
Fixes #936 

Changes:Allowed editing of events

### CODE REVIEW

File: app/src/main/java/org/fossasia/openevent/app/data/event/Event.java [hunk 1]
```diff
@@ -113,4 +115,7 @@
 
     public Event() { }
 
+    @ColumnIgnore
+    @JsonIgnore
+    public ObservableBoolean selected = new ObservableBoolean();
```
[the remark this answers is not in the document]
Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/app/data/event/Event.java]: Avoid using ObservableBoolean and also avoid changing models

**Comment by sridharjajoo** [comment]
output_nnj3p8
**When the user long presses on any item :** 
1) If none of the other items are selected then the item on which user presses gets selected.
2) If an item is already selected then the other item gets selected and previous gets unselected.

**When the user just single press on any item:**
1) If some other item is already selected then all other gets unselected and fragment comes back from edit mode to normal mode.

@iamareebjamal Kindly review.

File: app/src/main/java/org/fossasia/openevent/app/data/event/Event.java [hunk 2]
```diff
@@ -105,12 +107,14 @@
     @JsonSerialize(using = ObservableStringSerializer.class)
     @JsonDeserialize(using = ObservableStringDeserializer.class)
     public ObservableString endsAt = new ObservableString();
+    @ColumnIgnore
+    @JsonIgnore
+    public ObservableBoolean selected = new ObservableBoolean();
```
Comment (Masquerade0097) [inline app/src/main/java/org/fossasia/openevent/app/data/event/Event.java]: @sridharjajoo As you can see in other PR's, we are avoiding the use of ObservableBoolean. So please use Boolean instead.

Comment (sridharjajoo) [reply app/src/main/java/org/fossasia/openevent/app/data/event/Event.java]: I tried changing Boolean from ObservableBoolean but the thing is that the highlighting in blue color which was happening earlier isn't happening now.

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/app/data/event/Event.java]: You can use ObservableBoolean, but out of the class and in the form of a Map inside presenter

Comment (sridharjajoo) [reply app/src/main/java/org/fossasia/openevent/app/data/event/Event.java]: @iamareebjamal Done the changes

**Comment by iamareebjamal** [comment]
A lot of code can be shared with create event components. There's a lot of duplication

**Comment by sridharjajoo** [comment]
@iamareebjamal I have moved the code from UpdateEventFragment, Presenter to the CreateFragment and Presenter. If this is fine, I'll subsequently remove the Update module in the next commit.

**Comment by iamareebjamal** [comment]
Nice work @sridharjajoo

**Comment by sridharjajoo** [comment]
@iamareebjamal I have removed the Update module.Kindly review

**Comment by iamareebjamal** [comment]
@sridharjajoo Please resolve conflicts

File: app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -18,7 +18,7 @@
         <provider
             android:name=".OrgaProvider"
             android:authorities="${applicationId}.org.fossasia.openevent.app"
-            android:exported="false"/>
+            android:exported="false" />
```
Comment (iamareebjamal) [inline app/src/main/AndroidManifest.xml]: Unnecessary change. Remove all changes from this file please

File: app/src/main/java/org/fossasia/openevent/app/common/di/module/android/ActivityBuildersModule.java [hunk 4]
```diff
@@ -34,5 +34,4 @@
 
     @ContributesAndroidInjector
     abstract ChartActivity contributeChartActivity();
-
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/common/di/module/android/ActivityBuildersModule.java]: This too

File: app/src/main/java/org/fossasia/openevent/app/core/event/create/CreateEventActivity.java [hunk 5]
```diff
@@ -24,10 +24,16 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.create_event_activity);
 
-        if (savedInstanceState == null) {
+        long id = getIntent().getLongExtra("eventid", -1);
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/core/event/create/CreateEventActivity.java]: Extract constant "eventid"

File: app/src/main/java/org/fossasia/openevent/app/core/event/create/CreateEventFragment.java [hunk 6]
```diff
@@ -79,10 +94,17 @@ public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
 
         setHasOptionsMenu(true);
 
-        binding.submit.setOnClickListener(view -> {
-            if (validator.validate())
-                getPresenter().createEvent();
-        });
+        if(eventId != -1) {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/core/event/create/CreateEventFragment.java]: Move the if logic inside the click listener

Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/core/event/create/CreateEventFragment.java]: Also, `eventId != -1` is checked repeatedly. It'd be better if this is changed to a logical boolean property like `isUpdateEvent` or better

File: app/src/main/java/org/fossasia/openevent/app/core/event/list/EventListFragment.java [hunk 7]
```diff
@@ -96,6 +99,20 @@ public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
...
+        super.onPrepareOptionsMenu(menu);
+        MenuItem menuItemSort = menu.findItem(R.id.sort);
+        MenuItem menuItemEdit = menu.findItem(R.id.edit);
+        if (editMode) {
+            menuItemSort.setVisible(false);
+            menuItemEdit.setVisible(true);
+        } else {
+            menuItemSort.setVisible(true);
+            menuItemEdit.setVisible(false);
+        }
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/core/event/list/EventListFragment.java]: Can be reduced to

```Java
menuItemSort.setVisible(!editMode);
menuItemEdit.setVisible(editMode);
```

File: app/src/main/java/org/fossasia/openevent/app/core/event/list/EventListFragment.java [hunk 8]
```diff
@@ -105,7 +122,11 @@ public boolean onOptionsItemSelected(MenuItem item) {
             case R.id.sortByEventDate:
                 sortEvents(SORTBYDATE);
                 return true;
-            default:
+            case R.id.edit:
+                Intent intent = new Intent(getActivity(), CreateEventActivity.class);
+                intent.putExtra("eventid", id);
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/core/event/list/EventListFragment.java]: Use constant

File: app/src/main/java/org/fossasia/openevent/app/data/event/Event.java [hunk 9]
```diff
@@ -102,5 +102,4 @@
     public List<Ticket> tickets;
 
     public Event() { }
-
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/data/event/Event.java]: Unnecessary change

File: app/src/main/java/org/fossasia/openevent/app/core/event/create/CreateEventFragment.java (hunk 6 above)
Comment (sridharjajoo) [reply app/src/main/java/org/fossasia/openevent/app/core/event/create/CreateEventFragment.java]: Actually i am using it because in case when update event is chosen, then an event id is passed as a parameter with intent as it is needed.I can use usUpdatEvent boolean but i will need to putExtra( ) a boolean along with the event id

**Comment by sridharjajoo** [comment]
@iamareebjamal Done the required changes

**Comment by iamareebjamal** [comment]
You have to look at the Travis logs and see what's the error

File: app/src/main/java/org/fossasia/openevent/app/core/event/create/CreateEventFragment.java [hunk 10]
```diff
@@ -80,8 +95,12 @@ public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
         setHasOptionsMenu(true);
 
         binding.submit.setOnClickListener(view -> {
-            if (validator.validate())
-                getPresenter().createEvent();
+            if (validator.validate()) {
+                if (eventId != -1)
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/core/event/create/CreateEventFragment.java]: Change to logical boolean as suggested in previous review

File: app/src/main/java/org/fossasia/openevent/app/core/event/create/CreateEventFragment.java [hunk 11]
```diff
@@ -124,6 +143,13 @@ public void onStart() {
         validate(binding.form.externalEventUrlLayout, ValidateUtils::validateUrl, getResources().getString(R.string.url_validation_error));
         validate(binding.form.originalImageUrlLayout, ValidateUtils::validateUrl, getResources().getString(R.string.url_validation_error));
         validate(binding.form.paypalEmailLayout, ValidateUtils::validateEmail, getResources().getString(R.string.email_validation_error));
+        if(eventId != -1) {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/core/event/create/CreateEventFragment.java]: Same here and add space after if

File: app/src/main/java/org/fossasia/openevent/app/core/event/list/EventsPresenter.java [hunk 12]
```diff
@@ -75,4 +79,37 @@ public EventsView getView() {
...
+    public void unselectEvent(Event event) {
+        if (event != null && selectedMap.containsKey(event))
+            selectedMap.get(event).set(false);
+    }
+
+    public void toolbarEditMode(Event currentEvent) {
+        long id = 0;
+        if (!lastEvent.equals(currentEvent)) {
+            unselectEvent(lastEvent);
+        id = currentEvent.getId();
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/core/event/list/EventsPresenter.java]: Indentation error

File: app/src/main/java/org/fossasia/openevent/app/core/event/create/CreateEventFragment.java [hunk 13]
```diff
@@ -55,6 +55,9 @@
     private Validator validator;
     private ArrayAdapter<CharSequence> currencyAdapter;
     private ArrayAdapter<CharSequence> paymentCountryAdapter;
+    private static final String EVENT_ID = CreateEventActivity.EVENT_ID;
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/core/event/create/CreateEventFragment.java]: Why duplicating the field, it is public in the activity, use it as is

File: app/src/main/java/org/fossasia/openevent/app/core/event/list/EventListFragment.java [hunk 14]
```diff
@@ -63,6 +63,9 @@
 
     private Context context;
     private boolean initialized;
+    private boolean editMode;
+    private long id;
+    private static final String EVENT_ID = "event_id";
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/core/event/list/EventListFragment.java]: Same, use shared constant

**Comment by sridharjajoo** [comment]
Just a small doubt. Why shouldn't we use the EVENT_ID in a common Constant.java file as it is used for this purpose only?

**Comment by iamareebjamal** [comment]
Because it is local to a group of files. Using global constants is almost always a bad idea

**Comment by sridharjajoo** [comment]
I have seen that for Update Sponsors as well as Update Tracks, there is code duplication. Should I try to handle them as well in a single module?

**Comment by iamareebjamal** [comment]
Yes, create an issue please
